File: app/src/test/java/com/ndmsystems/coala/layers/arq/InMemoryDataTest.java
package com.ndmsystems.coala.layers.arq;

import com.ndmsystems.coala.layers.arq.data.InMemoryData;

import org.junit.Test;

import static org.junit.Assert.*;

/**
 * Created by Владимир on 16.08.2017.
 */
public class InMemoryDataTest {

    @Test
    public void onAppend_1to2_shouldResultWith12() {
        InMemoryData data1 = new InMemoryData(new byte[]{1});
        InMemoryData data2 = new InMemoryData(new byte[]{2});

        data1.append(data2);

        byte[] bytes = data1.get();
        assertEquals(2, bytes.length);
        assertEquals(1, bytes[0]);
        assertEquals(2, bytes[1]);
    }

    @Test
    public void onGetRangeFrom3To5_shouldRsultWith34() {
        InMemoryData sourceData = new InMemoryData(new byte[]{0, 1, 2, 3, 4, 5, 6, 7, 8, 9});

        byte[] bytes = sourceData.get(3, 5);

        assertEquals(2, bytes.length);
        assertEquals(3, bytes[0]);
        assertEquals(4, bytes[1]);
    }

}

File: app/src/test/java/com/ndmsystems/coala/layers/arq/InMemoryDataTest.kt
package com.ndmsystems.coala.layers.arq

import com.ndmsystems.coala.layers.arq.data.IData
import com.ndmsystems.coala.layers.arq.data.InMemoryData
import org.spekframework.spek2.Spek
import org.spekframework.spek2.style.gherkin.Feature
import kotlin.test.assertEquals


/*
 * Created by Evgenii Stepanov on 04.09.19
 */

object InMemoryDataTest : Spek({

    Feature("Testing InMemoryData") {

        Scenario("append one IData to another") {

            lateinit var dataOne: IData
            lateinit var dataTwo: IData

            Given("two InMemory data objects") {
                dataOne = InMemoryData(byteArrayOf(1))
                dataTwo = InMemoryData(byteArrayOf(2))
            }

            When("append values of second to first") {
                dataOne.append(dataTwo)
            }

            Then("size should be equal 2") {
                assertEquals(2, dataOne.size())
            }

            Then("first should contains all elements") {
                assertEquals(1.toByte(), dataOne.get()[0])
                assertEquals(2.toByte(), dataOne.get()[1])
            }

        }

        Scenario("extract arbitrary range of bytes") {
            lateinit var data: IData
            lateinit var result: ByteArray

            Given("non-emtpty InMemoryData object") {
                data = InMemoryData(byteArrayOf(0, 1, 2, 3, 4, 5, 6, 7, 8, 9))
            }

            When("get bytes between range 4..7") {
                result = data.get(4, 7)!!
            }

            Then("size should be equal 3") {
                assertEquals(3, result.size)
            }

            Then("dataOne should contains all elements") {
                assertEquals(4.toByte(), result[0])
                assertEquals(5.toByte(), result[1])
                assertEquals(6.toByte(), result[2])
            }
        }
    }
})